core/src/main/java/com/uber/m3/tally/ScopeImpl.java

```diff
@@ -178,24 +127,17 @@ public void close() {
         }
     }
 
+    <T extends Reportable> void addToReportingQueue(T metric) {
+        reportingQueue.add(metric);
+    }
+
     /**
      * Reports using the specified reporter.
      * @param reporter the reporter to report
      */
     void report(StatsReporter reporter) {
-        for (Map.Entry<String, CounterImpl> counter : counters.entrySet()) {
-            counter.getValue().report(fullyQualifiedName(counter.getKey()), tags, reporter);
-        }
-
-        for (Map.Entry<String, GaugeImpl> gauge : gauges.entrySet()) {
-            gauge.getValue().report(fullyQualifiedName(gauge.getKey()), tags, reporter);
-        }
-
-        // No operations on timers required here; they report directly to the StatsReporter
-        // i.e. they are not buffered
-
-        for (Map.Entry<String, HistogramImpl> histogram : histograms.entrySet()) {
-            histogram.getValue().report(fullyQualifiedName(histogram.getKey()), tags, reporter);
+        for (Reportable metric : reportingQueue) {
```

Comment:
Now I have gone thru this PR, I still don't understand this comment https://github.com/uber-java/tally/pull/73#discussion_r459820758, the traversal scope isn't reduced, it is still all the metrics. And whether to report or not is decided by the `report` call on each metric, which is the same prior to this refactor. What am I missing here?

Reply:
It's not: previously for histograms we're traversing all the available buckets no matter whether their counters had changed or not. Having moderately large number of counters (~100) causes substantial CPU churn of simply looping through them multiplied by the fact that we have thousands of them.

Reply:
This diff flips this process -- each histogram bucket is created lazily only when it's counter is about to change, and its bucket will individually add itself to the reporting queue when created.

Reply:
This yields substantial reduction of CPU churn for low-percentages of buckets being used (~4.5x). It's captured in the benchmark results in the description.

Reply:
Got it. I see the difference now. Another optimization we can do is to improve reporting for all metric types, current reporting approach is to iterate through all metrics and call `report` on each one, at which time whether to report or not is checked. We can instead maintain a dynamic reporting queue where the metrics in the queue always have changes to report, i.e, enqueue on change and dequeue after reporting. Obviously that's something we can do in a separate PR.

Reply:
Yeah, that was the long-term plan -- to actually use dynamic reporting queue